File: lib/dart/lib/src/transport/nats_transport.dart
part of frugal.transport;

class NatsTransportFactory implements TransportFactory {
  Nats client;

  NatsTransportFactory(this.client);

  Transport getTransport() => new NatsTransport(this.client);
}


class NatsTransport implements Transport {
  TTransport tTransport;
  NatsThriftTransport nTransport;

  NatsTransport(Nats client) {
    var tr = new NatsThriftTransport(client);
    tTransport = tr;
    nTransport = tr;
  }

  Future subscribe(String subject) {
    nTransport.setSubject(subject);
    return nTransport.open();
  }

  Future unsubscribe() {
    return nTransport.close();
  }

  void preparePublish(String subject) {
    nTransport.setSubject(subject);
  }

  TTransport thriftTransport() => tTransport;

  Future applyProxy(TTransportFactory transportFactory) async {
    tTransport = await transportFactory.getTransport(tTransport);
  }
}

File: lib/java/src/main/java/com/workiva/frugal/NatsTransport.java
package com.workiva.frugal;

import org.apache.thrift.transport.TTransport;
import org.apache.thrift.transport.TTransportException;
import org.apache.thrift.transport.TTransportFactory;

/**
 * NatsTransport is an implementation of the Transport interface backed by the NATS
 * messaging system.
 */
public class NatsTransport implements Transport {

    private TTransport thriftTransport;
    private NatsThriftTransport nats;

    @Override
    public void subscribe(String topic) throws TTransportException {
        nats.setSubject(topic);
        thriftTransport.open();
    }

    @Override
    public void unsubscribe() throws TTransportException {
        thriftTransport.close();
    }

    @Override
    public void preparePublish(String topic) {
        nats.setSubject(topic);
    }

    @Override
    public TTransport thriftTransport() {
        return thriftTransport;
    }

    @Override
    public void applyProxy(TTransportFactory proxy) {
        thriftTransport = proxy.getTransport(thriftTransport);
    }

}
